Part 2 of Standalone Components

alxhub opened the pull request:
Part 1 implemented the `standalone` flag in components/directives/pipes, and `imports` for components. In this PR, NgModules gain support for importing standalone types.

alxhub committed:
refactor(compiler-cli): remove NgModule tracking from scope data

Before the `SemanticSymbol` system which now powers incremental compilation,
the compiler previously needed to track which NgModules contributed to the
scope of a component in order to recompile correctly if something changed.
This commit removes that legacy field (which had no consumers) as well as the
logic to populate it.

alxhub committed:
fix(compiler-cli): better error messages for NgModule structural issues

This commit improves the error messages generated by the compiler when NgModule
scope analysis finds structural issues within a compilation. In particular,
errors are now shown on a node within the metadata of the NgModule which
produced the error, as opposed to the node of the erroneous declaration/import/
export. For example, if an NgModule declares `declarations: [FooCmp]` and
`FooCmp` is not annotated as a directive, component, or pipe, the error is now
shown on the reference to `FooCmp` in the `declarations` array expression.
Previously, the error would have been shown on `FooCmp` itself, with a mention
in the error text of the NgModule name.

Additional error context in some cases has been moved to related information
attached to the diagnostic, which further improves the legibility of such
errors. Error text has also been adjusted to be more succinct, since more info
about the error is now delivered through context.

alxhub committed:
test(compiler-cli): split standalone tests into component & module tests

This just helps organize the standalone tests a little bit.

alxhub committed:
feat(compiler-cli): standalone types imported into NgModule scopes

This commit implements the next step of Angular's "standalone" functionality,
by allowing directives/components/pipes declared as `standalone` to be imported
into NgModules. Errors are raised when such a type is not standalone but is
included in an NgModule's imports.

alxhub committed:
refactor(compiler-cli): move error for declaring standalone things

This commit moves the error for declaring a standalone directive/component/
pipe to the `LocalModuleScopeRegistry`. Previously the error was produced
by the `NgModuleHandler` directly.

Producing the error in the scope registry allows the scope to be marked as
poisoned when the error occurs, preventing spurious downstream errors from
surfacing.

pkozlowski-opensource commented on packages/compiler-cli/test/ngtsc/standalone_spec.ts:
```diff
@@ -264,6 +266,54 @@ runInEachFileSystem(() => {
+      
+        @NgModule({
+          declarations: [TestCmp],
+          imports: [StandaloneCmp],
```
Maybe add a test for importing standalone directives and pipes as well?

krishna96-tecdune replied:
test

pkozlowski-opensource commented on packages/compiler-cli/test/ngtsc/standalone_spec.ts:
```diff
@@ -264,6 +266,54 @@ runInEachFileSystem(() => {
+          export class TestDir {}
+      
+          @NgModule({
+            imports: [TestDir],
```
Maybe add a test for importing a non-standalone component and pipes as well?

atscott commented on packages/compiler-cli/src/ngtsc/scope/src/util.ts:
```diff
@@ -0,0 +1,65 @@
+    // current compilation, but not those imported from .d.ts dependencies. We could likely scan the
+    // program here and find NgModules to suggest, to improve the error in that case.
+  }
+  let extraText: string = '';
+  if (relatedInformation === undefined) {
+    // If no contextual pointers can be provided to suggest a specific remedy, then at least tell
+    // the user broadly what they need to do.
+    extraText = ' It must be imported via an NgModule.';
+  }
+  return makeDiagnostic(
+      ErrorCode.COMPONENT_IMPORT_NOT_STANDALONE, getDiagnosticNode(ref, rawExpr),
+      `The ${kind} '${
+          ref.node.name
+              .text}' appears in 'imports', but is not standalone and cannot be imported directly.${
+          extraText}`,
```
```suggestion
  let diagnosticMessage = `The ${kind} '${
          ref.node.name
              .text}' appears in 'imports', but is not standalone and cannot be imported directly`;
  if (relatedInformation === undefined) {
    // If no contextual pointers can be provided to suggest a specific remedy, then at least tell
    // the user broadly what they need to do.
    diagnosticMessage += ' It must be imported via an NgModule.';
  }
  return makeDiagnostic(
      ErrorCode.COMPONENT_IMPORT_NOT_STANDALONE, getDiagnosticNode(ref, rawExpr),
      diagnosticMessage,
```

Similar to the other comment, I think maybe this is a tad more straightforward.

alxhub replied:
👍 much cleaner.

atscott commented on packages/compiler-cli/src/ngtsc/scope/src/local.ts:
```diff
@@ -623,10 +651,27 @@ function invalidTransitiveNgModuleRef(
   return makeDiagnostic(
       ErrorCode.NGMODULE_INVALID_REEXPORT, getDiagnosticNode(decl, rawExpr),
-      `Can't be exported from this NgModule as it's neither declared nor imported.`);
+      `Can't be exported from this NgModule as ${specificIssue}.`);
```
Nit: The construction of the sentence reads a bit weird when it's split into two variables like this. Especially since when you read the code, you're reading the end of the sentence first (without a period), and then you later read the beginning of the sentence. It might be helpful to just write it as two separate sentences.

```
let diagnosticMessage = `Can't be exported from this NgModule.`
if (x) {
  diagnosticMessage += ` It must be imported first.`;
} else if (y) {
 ...
} else {
 ...
}
return makeDiagnostic(x, y, diagnosticMessage);
```

alxhub replied:
👍 done

atscott commented on packages/compiler-cli/src/ngtsc/testing/src/utils.ts:
```diff
@@ -147,3 +148,16 @@ export function getSourceCodeForDiagnostic(diag: ts.Diagnostic): string {
+    throw new Error(`Expected ts.Diagnostic to have a file source`);
+  }
+  const node = getTokenAtPosition(diag.file, diag.start!);
+  debugger;
```
nit: debugger;

atscott commented on packages/compiler-cli/src/ngtsc/testing/src/utils.ts:
```diff
@@ -147,3 +148,16 @@ export function getSourceCodeForDiagnostic(diag: ts.Diagnostic): string {
+    guard: (node: ts.Node) => node is T): T {
+  const diag = diagnostic as ts.Diagnostic | ts.DiagnosticRelatedInformation;
+  if (diag.file === undefined) {
+    throw new Error(`Expected ts.Diagnostic to have a file source`);
```
```suggestion
    throw new Error(`Expected ts.Diagnostic to have a file source.`);
```
nit

alxhub replied:
This is actually existing code, just moved.

atscott reviewed (dismissed):
As mentioned by Pawel, this could generally use more test coverage for the various diagnostics and types of standalone entities.

alxhub committed:
feat(compiler-cli): propagate `standalone` flag to runtime

This commit carries the `standalone` flag forward from a directive/pipe
into its generated directive/pipe definition, allowing the runtime to
recognize standalone entities.

alxhub commented on packages/compiler-cli/src/ngtsc/scope/src/local.ts:
```diff
@@ -308,11 +308,33 @@ export class LocalModuleScopeRegistry implements MetadataRegistry, ComponentScop
       const directive = this.localReader.getDirectiveMetadata(decl);
       const pipe = this.localReader.getPipeMetadata(decl);
       if (directive !== null) {
+        if (directive.isStandalone) {
+          const refType = directive.isComponent ? 'Component' : 'Directive';
+          diagnostics.push(makeDiagnostic(
+              ErrorCode.NGMODULE_DECLARATION_IS_STANDALONE,
+              decl.getOriginForDiagnostics(ngModule.rawDeclarations!),
+              `${refType} ${
+                  decl.node.name
+                      .text} is standalone, and cannot be declared in an NgModule. Did you mean to import it instead?`));
+          isPoisoned = true;
+          continue;
+        }
```
👍 we have a test for this!

JoostK commented on packages/compiler-cli/src/ngtsc/scope/src/local.ts:
```diff
@@ -390,9 +417,11 @@ export class LocalModuleScopeRegistry implements MetadataRegistry, ComponentScop
+        const dirMeta = this.fullReader.getDirectiveMetadata(decl);
+        const pipeMeta = this.fullReader.getPipeMetadata(decl);
+        if (dirMeta !== null || pipeMeta !== null) {
+          const isStandalone = (dirMeta?.isStandalone || pipeMeta?.isStandalone)!;
```
nit: a `??` would be slightly more appropriate here, also allowing to remove one `?.` in favor of `!`.

```suggestion
          const isStandalone = dirMeta?.isStandalone ?? pipeMeta!.isStandalone;
```

alxhub replied:
I think this is too magical actually (both the original and the suggested) - I've changed it to be more explicit.

JoostK commented on packages/compiler-cli/src/ngtsc/scope/src/util.ts:
```diff
@@ -0,0 +1,64 @@
+  if (scope !== null) {
+    // The directive/pipe in question is declared in an NgModule. Check if it's also exported.
+    const isExported = scope.exported.directives.some(exp => exp.ref.node === ref.node) ||
+        scope.exported.pipes.some(exp => exp.ref.node);
```
This is missing `=== ref.node` in the search among exported pipes.

JoostK commented on packages/compiler-cli/src/ngtsc/scope/src/local.ts:
```diff
@@ -623,10 +642,26 @@ function invalidTransitiveNgModuleRef(
+  let message = `Can't be exported from this NgModule, as `;
+  if (isStandalone) {
+    // Standalone types need to be imported into an NgModule before they can be re-exported.
+    message += 'it must be imported first';
```
I don't see tests for the case where a standalone entity is exported from an NgModule that doesn't import it.

JoostK commented on packages/core/src/render3/definition.ts:
```diff
@@ -284,6 +284,11 @@ export function ɵɵdefineComponent<T>(componentDefinition: {
+  /**
+   * Whether this directive/component is standalone.
+   */
+  isStandalone?: boolean;
```
Since this is optional, can we do `isStandalone?: true;` here or would that be inaccurate?

JoostK commented on packages/core/src/render3/definition.ts:
```diff
@@ -312,6 +317,7 @@ export function ɵɵdefineComponent<T>(componentDefinition: {
       onPush: componentDefinition.changeDetection === ChangeDetectionStrategy.OnPush,
       directiveDefs: null!,  // assigned in noSideEffects
       pipeDefs: null!,       // assigned in noSideEffects
+      isStandalone: componentDefinition.isStandalone !== true,
```
The `!== true` seems like a bug here?!

pkozlowski-opensource replied:
Looks like it to me as well. Also points out that we are probably missing a test for this case (not sure what would be the best test case, though, as we are not using this runtime flag anywhere yet).

alxhub replied:
Yeah, I expect we'll add tests once we have the full runtime output defined.

JoostK commented on packages/core/src/render3/definition.ts:
```diff
@@ -709,13 +715,19 @@ export function ɵɵdefinePipe<T>(pipeDef: {
     name: pipeDef.name,
     factory: null,
     pure: pipeDef.pure !== false,
+    isStandalone: pipeDef.isStandalone !== true,
```
The `!== true` seems like a bug here?!

alxhub commented on packages/compiler-cli/test/ngtsc/ngtsc_spec.ts:
```diff
@@ -5291,7 +5291,10 @@ function allTests(os: string) {
         expect(errors.length).toBe(1);
-        const {code, messageText} = errors[0];
+        let {code, messageText} = errors[0];
+        if (errors[0].relatedInformation !== undefined) {
+          messageText += errors[0].relatedInformation.map(info => info.messageText).join('\n');
+        }
```
I don't think so, because `relatedInformation` is not the same as using a `DiagnosticMessageChain` for `messageText`. But I might be misunderstanding the TS APIs here?

JoostK replied:
Oh, you are correct. It could be used within the `.map(info => ...)` transform here to avoid an `[object object]` string in the case of a message chain, but we can adjust once we need it (since this is just testing code).

ahasall commented on packages/core/src/render3/interfaces/definition.ts:
```diff
@@ -355,6 +360,11 @@ export interface PipeDef<T> {
+  /**
+   * Whether this pipe is standalone.
+   */
+  isStandalone: boolean;
```
Is it possible to rename `isStandalone` to just `standalone` ? It’s shorter and consistent with what we already have with the `pure` option of pipes.

pkozlowski-opensource reviewed (approved):
LGTM

dylhunn commented:
This PR was merged into the repository by commit 2142ffd295de491eb8582c2eb1712b5b48044f24.
